### DISCUSSION TITLE
iTwin on the Hololens 2

### DISCUSSION THREAD

**Discussion by WillProicere** [iTwin/itwinjs-core#5046] [Q&A]
I have a project built for the Hololens 2 that has a model I imported from the iTwin application. It's using the combined mesh. When I launch the project from unreal in the VR preview in the headset everything works as expected, however, when packaging the project and loading it onto the Hololens 2, when I try to run it it shows the application loading and then quits.

Using UE 5.1.0.

I've tried several different ways to fix the issues and the closest I've come is setting the mesh from the combinedmodel to none, this way I saw the UI that I created.

Things that didn't work: setting the material to opaque, getting a smaller section of the scene from iTwin, making the mesh smaller in unreal. Each of these attempts led to the same outcome of loading, into quitting.

**Comment by MichaelBelousov** [comment]
@WillProicere by `iTwin application` you mean the [`iTwin Exporter for Datasmith`](https://qa-developer.bentley.com/unreal/integration/overview/), correct? I assume yes since you refer to its "combined mesh" option.

Can you run the packaged project under the visual studio debugger? You should need visual studio to build the iTwin Datasmith plugin for a packaged UE4 project anyway. This should give you more information on what's going wrong in the "output" window of visual studio.

**Reply by WillProicere** [reply]
@MichaelBelousov Sorry, yes it is the iTwin Exporter for Datasmith plugin.

I did find a workaround which was to build the application for Windows and then stream it to the Hololens, this way works but is fairly choppy so I would like to find a way to run it from the headset.

Running the application packaged for Windows through VS, the only error I got was :
"LogJson: Error: Json Value of type 'Null' used as a 'Array'.
LogJson: Warning: Field transformTimeline was not found."

This was listed a lot but not sure what it means, I also got it when running in the UE editor but didn't think anything of it since it ran.

**Reply by MichaelBelousov** [reply]
In your project you should have a file from when you installed our plugin, `Plugins/ITwinUnrealDatasmith/Source/ITwinUnrealDatasmithPlugin/Private/ScheduledActor.cpp`
Line 121 of that file is where this occurs, it assumes that the schedule script stored in your datasmith metadata contains a transform timeline for every animation batch.

You can replace that line 
```cpp
for (const auto& KeyFrameVal : ElementTimeline->GetArrayField(FString("transformTimeline")))
{
```
with

```cpp
TArray<TSharedPtr<FJsonValue>> outTransformTimelineJson;
ElementTimeline->TryGetArrayField(FString("transformTimeline"), &outTransformTimelineJson);
for (const auto& KeyFrameVal : outTransformTimelineJson)
{
```

And rebuild and retry. As for why the windows platform doesn't crash when encountering this, I don't know. If this works for you, we will patch the plugin to include that fix.

**Reply by WillProicere** [reply]
Getting an error on this line "ElementTimeline->TryGetArrayField(FString("transformTimeline"), &outTransformTimelineJson);"

error C2664: 'bool FJsonObject::TryGetArrayField(const FString &,const TArray<TSharedPtr<FJsonValue,ESPMode::ThreadSafe>,FDefaultAllocator> *&) const': cannot convert argument 2 from 'TArray<TSharedPtr<FJsonValue,ESPMode::ThreadSafe>,FDefaultAllocator> *' to 'const TArray<TSharedPtr<FJsonValue,ESPMode::ThreadSafe>,FDefaultAllocator> *&'

**Reply by MichaelBelousov** [reply]
Sorry @WillProicere . I booted up my windows machine and confirmed this should work:

```cpp
using JsonArray = TArray<TSharedPtr<FJsonValue>>;
const JsonArray* outTransformTimelineJson;
ElementTimeline->TryGetArrayField(FString("transformTimeline"), outTransformTimelineJson);
for (const auto& KeyFrameVal : outTransformTimelineJson != nullptr ? *outTransformTimelineJson : JsonArray())
{
```

If after that you still get similar errors about `visibilityTimeline` or `colorTimeline` or `cuttingPlaneTimeline`, you will need to apply the same fix to the similar looking following lines (after patch) `133`, `148` and `172`, but renaming `outTransformTimelineJson` to something like `outColorTimelineJson`, and dropping the `using JsonArray =` line. Please let me know what results and if you need help applying further patches. I would also be interested in you [emailing me](mailto:mike.belousov@bentley.com) your `.udatasmith` file so I can see what it's missing and adjust the plugin accordingly.

**Reply by WillProicere** [reply]
That fixed the issue of not being able to build it, I didn't need to change anything else. 
However unfortunately once packaged and downloaded onto the Hololens the results end up being the same, I see the four loading spheres and then it quits out of the application

**Reply by MichaelBelousov** [reply]
@WillProicere No errors reported in logs?

**Reply by MichaelBelousov** [reply]
@WillProicere can you run your application via visual studio to see logs and exceptions thrown?

**Reply by WillProicere** [reply]
Sorry was out of office for the weekend, running the packaged debug version and I saw the same "LogJson: Error: Json Value of type 'Null' used as a 'Array'. error for the other _____Timeline variables so after applying the same fix as for transformTimeline I can now build it from VS but running it in the editor it triggers a breakpoint on this line:

for (const auto& KeyFrameVal : outTransformTimelineJson != nullptr ? *outTransformTimelineJson : JsonArray())

This wasn't the case when the only code changed was on the transformTimeline for loop.

**Reply by MichaelBelousov** [reply]
Does it give a reason for triggering a breakpoint? Since my small patches haven't been sufficient, please [email me](mailto:mike.belousov@bentley.com) your export's .udatasmith file, and I will provide an updated plugin implementation that can handle any "null" fields in it.

**Reply by WillProicere** [reply]
![image](https://user-images.githubusercontent.com/118467336/222094702-84c42983-d953-44d8-98a6-ad47f2e17b49.png)

This is the error I am getting.

Unfortunately I'm unable to share the file as I don't have permission.

**Reply by MichaelBelousov** [reply]
@WillProicere my bad please adjust the patch to default the timeline pointer to null:
```cpp
const JsonArray* outTransformTimelineJson = nullptr;
```
That `0xFFFFFFFFFFF...` thing means I forgot to initialize that field and it's not technically null but still invalid

**Reply by MichaelBelousov** [reply]
As for not being able to share the data, would you be willing to take an HTML file I write and send you here, which executes some javascript in your browser, running fully locally on your machine without sending your data anywhere (you can disable network if you'd like to be sure), which will let you drag and drop in the .udatasmith file, and it will discover the "shape" of the data and report it to you crudely, which you can copy back into this chat for me?

**Reply by MichaelBelousov** [reply]
I have released these patches as an update.
